feat(npm/runtime): add `decorateMetadata` helper function

Dunqing opened the pull request:
Copy from https://github.com/microsoft/TypeScript/blob/d85767abfd83880cea17cea70f9913e9c4496dcc/src/compiler/factory/emitHelpers.ts#L744-L753

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9189?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9190** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9190?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9189** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9189?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9189?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9057** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9057?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9185** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9185?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9189).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel reviewed (commented):
`Reflect.metadata` is not a native JS API. According to TS docs, it requires NPM package https://www.npmjs.com/package/reflect-metadata to "polyfill" this API.

Just checking that we don't need to do anything on our side to insert `import 'reflect-metadata';` or something?

I assume the idea is that the user would handle this themselves, but I'm not up to speed on this whole decorator metadata thing, so I just wanted to check.

Dunqing commented:
> `Reflect.metadata` is not a native JS API. According to TS docs, it requires NPM package https://www.npmjs.com/package/reflect-metadata to "polyfill" this API.
> 
> Just checking that we don't need to do anything on our side to insert `import 'reflect-metadata';` or something?
> 
> I assume the idea is that the user would handle this themselves, but I'm not up to speed on this whole decorator metadata thing, so I just wanted to check.

Yes, we don't need to do anything with `reflect-metadata`, this usually proceeds by DI Framework like `NestJS`. NestJS [documentation](https://docs.nestjs.com/#alternatives) mentioned, that if you want to create a NestJS APP from scratch, you need to manually add a few dependencies one of them is `reflect-metadata`.

overlookmotel commented:
Great. Thanks for explaining.

Dunqing committed:
feat(npm/runtime): add `decorateMetadata` helper function (#9189)

Copy from https://github.com/microsoft/TypeScript/blob/d85767abfd83880cea17cea70f9913e9c4496dcc/src/compiler/factory/emitHelpers.ts#L744-L753
